go-controller/pkg/ovn/controller/services/repair.go

```diff
@@ -0,0 +1,113 @@
+	serviceTracker *serviceTracker
+}
+
+// NewRepair creates a controller that periodically ensures that there is no stale data in OVN
+func NewRepair(interval time.Duration,
+	serviceTracker *serviceTracker,
+) *Repair {
+	return &Repair{
+		interval:       interval,
+		serviceTracker: serviceTracker,
+	}
+}
+
+// RunUntil starts the controller until the provided ch is closed.
+func (r *Repair) RunUntil(stopCh <-chan struct{}) {
+	wait.Until(func() {
+		if err := r.RunOnce(); err != nil {
+			klog.Errorf("Error repairing services: %v")
+		}
+	}, r.interval, stopCh)
+}
+
+// RunOnce verifies the state of Services and OVN LBs and returns an error if an unrecoverable problem occurs.
+func (r *Repair) RunOnce() error {
+	return retry.RetryOnConflict(retry.DefaultBackoff, r.runOnce)
+}
+
+// runOnce verifies the state of the cluster OVN LB VIP allocations and returns an error if an unrecoverable problem occurs.
+func (r *Repair) runOnce() error {
+	startTime := time.Now()
+	klog.V(4).Infof("Starting repairing loop for services")
+	defer func() {
+		klog.V(4).Infof("Finished repairing loop for services: %v", time.Since(startTime))
+		metrics.MetricSyncServiceLatency.WithLabelValues("repair-loop").Observe(time.Since(startTime).Seconds())
+	}()
+
+	// Obtain all the load balancers UUID
+	ovnLBCache := make(map[v1.Protocol][]string)
+	// We have different loadbalancers per protocol
+	protocols := []v1.Protocol{v1.ProtocolSCTP, v1.ProtocolTCP, v1.ProtocolUDP}
+	// ClusterIP OVN load balancers
+	for _, p := range protocols {
+		lbUUID, err := loadbalancer.GetOVNKubeLoadBalancer(p)
+		if err != nil {
+			return errors.Wrapf(err, "Failed to get OVN load balancer for protocol %s", p)
+		}
+		ovnLBCache[p] = append(ovnLBCache[p], lbUUID)
+	}
+	// NodePort, ExternalIPs and Ingress OVN load balancers
+	gatewayRouters, _, err := gateway.GetOvnGateways()
+	if err != nil {
+		return errors.Wrapf(err, "Failed to get gateway router")
+	}
+	for _, p := range protocols {
+		for _, gatewayRouter := range gatewayRouters {
+			lbUUID, err := gateway.GetGatewayLoadBalancer(gatewayRouter, p)
+			if err != nil {
+				return errors.Wrapf(err, "Failed to get OVN GR load balancer for protocol %s", p)
+			}
+			ovnLBCache[p] = append(ovnLBCache[p], lbUUID)
+		}
+	}
+
+	// Get Kubernetes Service state
+	svcVipsProtocolMap := r.serviceTracker.getServiceVipsMap()
+
+	// Reconcile with OVN state
+	// Obtain all the VIPs present in the OVN LoadBalancers
+	// and delete the ones that are not present in the service tracker
+	for _, p := range protocols {
+		for _, lb := range ovnLBCache[p] {
+			vips, err := loadbalancer.GetLoadBalancerVIPs(lb)
+			if err != nil {
+				return errors.Wrapf(err, "Failed to get load balancer vips for %s", ovnLBCache[p])
+			}
+			for vip := range vips {
+				key := virtualIPKey(vip, p)
+				// Virtual IP and protocol doesn't belong to a Kubernetes service
+				if !svcVipsProtocolMap.Has(key) {
```

Comment:
Should we log that a stale entry has been found and that it is being deleted? I think that would be helpful for debugging in case something is deleted unintentionally

Reply:
absolutely thanks, I thought I was already doing it